### PR TITLE
PID-231: Refactor OPAC Learner

### PR DESCRIPTION
**Pull request by gauthamvasan**

### CODE REVIEW

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 1]
```diff
@@ -1,161 +1,210 @@
...
+        self.indices = indices
+
+    def __iter__(self):
+        return iter(self.indices)
 
+    def __len__(self):
+        return len(self.indices)
+
+
+class ExperienceChunk(object):
```
Comment (jaberg) [inline rlscan_train/agent/opaclearner_hdf5.py]: suggest to rename this class... it's now an iterator over the experience chunks?

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 2]
```diff
@@ -1,161 +1,210 @@
...
-                    total_loss = 0
-
-
-            # TODO: Final update should be encoded in for loop
-            total_norm = torch.nn.utils.clip_grad_norm(self.actor_critic.parameters(), self.cfg['max_grad_clip'])
-            print("Grad norm: ", total_norm, "Total loss: ", total_loss)
-            self.actor_critic_opt.step()
+                preu_vals_chunk, preu_lprobs_chunk = self.actor_critic.val_lprob(obss_chunk, acts_chunk)
+
+            rews = torch.cat((rews, rews_chunk), dim=0)
```
Comment (jaberg) [inline rlscan_train/agent/opaclearner_hdf5.py]: The old list approach was better I think, this one is an O(N^2) amount of copying into bigger and bigger temporaries.

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 3]
```diff
@@ -1,161 +1,210 @@
...
+                    if policy_surr is None:
+                        print("Stop aggregating gradients to prevent over-fitting")
+                        break
+
+                    policy_surr /= num_blocks
+                    total_loss += policy_surr
+                    policy_surr.backward()
+
+                total_norm = torch.nn.utils.clip_grad_norm(self.actor_critic.parameters(), self.cfg['max_grad_clip'])
+                print("Grad norm: ", total_norm, "Total loss: ", total_loss)
```
Comment (jaberg) [inline rlscan_train/agent/opaclearner_hdf5.py]: are you switching over to logging yet?

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 4]
```diff
@@ -1,161 +1,210 @@
...
+                for gpu_block_indices in BatchSampler(sampler=SubsetSampler(indices=opt_batch_indices),
+                                                      batch_size=self.GPU_BLOCK_SIZE, drop_last=False):
+                    obss, acts, _, _, _ = self.buffer[experience_chunk.start_ind + np.array(gpu_block_indices)]
+                    new_vals, new_lprobs = self.actor_critic.val_lprob(obss, acts)
+                    policy_surr = self.actor_critic_update(advs[gpu_block_indices], data_lprobs[gpu_block_indices],
+                                                           new_lprobs, preu_lprobs[gpu_block_indices], annealer,
+                                                           rets[gpu_block_indices], new_vals,
+                                                           preu_vals[gpu_block_indices])
+                    if policy_surr is None:
+                        print("Stop aggregating gradients to prevent over-fitting")
```
Comment (jaberg) [inline rlscan_train/agent/opaclearner_hdf5.py]: logging?

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 5]
```diff
@@ -166,71 +215,92 @@ def actor_critic_update(self, advs, data_lprobs, new_lprobs, preu_lprobs, anneal
...
+
+        lam = self.cfg['lambda'] * torch.min(1 / (ratio + np.finfo(float).eps), torch.Tensor([1]))
         for i in reversed(range(rews.size(0))):
             deltas[i] = rews[i] + self.cfg['gamma'] * prev_vals * (1 - dones[i]) - vals[i]
-            lam = self.cfg['lambda'] * torch.min(1 / (ratio[i] + np.finfo(float).eps), torch.Tensor([1]))
-            advs[i] = ratio[i] * (deltas[i] + self.cfg['gamma'] * lam * prev_adv * (1 - dones[i]))
-
-            prev_vals = vals[i, 0]
-            prev_adv = advs[i, 0]
+            advs[i] = ratio[i] * (deltas[i] + self.cfg['gamma'] * lam[i] * prev_adv * (1 - dones[i]))
```
Comment (jaberg) [inline rlscan_train/agent/opaclearner_hdf5.py]: is there an off-by-one error here?

Comment (jaberg) [reply rlscan_train/agent/opaclearner_hdf5.py]: (I mean in the range of values of `i`)

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 6]
```diff
@@ -166,71 +215,92 @@ def actor_critic_update(self, advs, data_lprobs, new_lprobs, preu_lprobs, anneal
...
 
         if self.cfg['norm_advantage']:
             advs = (advs - advs.mean())
-            if advs.std() != 0 and not torch.isnan(advs.std()):
-                advs /= advs.std()
-
-        # TODO: Update running statistics using rets
+            std = advs.std()
+            if std < 1e-3:  # XXX hyper-parameter, especially if we clip in future
+                print('Warning: estimate_off_advantage norm std() getting very small:', std)
```
Comment (jaberg) [inline rlscan_train/agent/opaclearner_hdf5.py]: print vs. logging

File: rlscan_train/agent/policies/scanac.py [hunk 7]
```diff
@@ -4,32 +4,33 @@
 from torch.utils.data import DataLoader
 from rlscan.policies.nnet_aug27_opac_discrete import ScanMotionPolicy_Aug27DiscreteScanAC
 
+GPU_INFERENCE_BATCH_SIZE = 20
```
Comment (jaberg) [inline rlscan_train/agent/policies/scanac.py]: start moving these to env variables?

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 8]
```diff
@@ -1,161 +1,210 @@
...
+"""
+N.B: These constants were chosen based on the computational hardware available on Vader:
+    - Processor: Intel® Core™ i7-4790K CPU @ 4.00GHz × 8
+    - RAM: 4 X 8192 DDR3 (32GB)
+    - GPU: 12GB Nvidia GeForce 1080Ti
+    
+    Please modify (if required) to better suit the compute available in your workstation 
+"""
+RAM_CHUNK_SIZE = 2000
+GPU_BLOCK_SIZE = 256
```
Comment (jaberg) [inline rlscan_train/agent/opaclearner_hdf5.py]: move to env variables?

File: rlscan_train/agent/opaclearner_hdf5.py (hunk 1 above)
Comment (gauthamvasan) [reply rlscan_train/agent/opaclearner_hdf5.py]: Right. I'm renamed it as `TrainingBatchLoader`.

File: rlscan_train/agent/opaclearner_hdf5.py (hunk 2 above)
Comment (gauthamvasan) [reply rlscan_train/agent/opaclearner_hdf5.py]: Good call! Reverting back to the list approach

File: rlscan_train/agent/opaclearner_hdf5.py (hunk 5 above)
Comment (gauthamvasan) [reply rlscan_train/agent/opaclearner_hdf5.py]: It's probably not clear from the code, but `batch_size == len(rews) == len(dones) == len(preu_vals) - 1 == len(data_lprobs) == len(preu_lprobs)`. The for loop considers length of `rews` and not `preu_vals`. This way, we get the right alignment.

File: rlscan_train/agent/buffer.py [hunk 9]
```diff
@@ -29,7 +33,7 @@ def __init__(self, random_state, memory_size, img_dim, filepath, hdf5_mode,
 
         # To ensure repeatability of learning updates
         np.random.set_state(random_state)
-        self.count = 0
+        self.buffer_len = 0
         self.memory_size = memory_size
```
Comment (chanb) [inline rlscan_train/agent/buffer.py]: I think we can replace `memory_size` with `max_buffer_len`. The variable names might be confusing to readers until they read further ahead.

File: rlscan_train/agent/buffer.py [hunk 10]
```diff
@@ -121,41 +132,42 @@ def sample_from_ind(self, indices):
...
             arm_sensory[arm_sensory != arm_sensory] = 0.
 
-        obss = torch.from_numpy(np.hstack((imgs, arm_sensory)).astype(np.float32))
-        acts = torch.from_numpy(np.stack(self.action_h5[indices, :]).astype(np.float32))
-        acts = acts.flatten() if self.action_dim == 1 else acts
+        obss = np.hstack((imgs, arm_sensory)).astype(np.float32)
+        acts = self.action_h5[index, :]
+        rews = self.reward_h5[index]
+        dones = self.done_h5[index].astype(np.float32)
+        lprobs = self.log_prob_h5[index]
```
Comment (chanb) [inline rlscan_train/agent/buffer.py]: I might be wrong, but I think if we use an `int` for `index`, we will have one less dimension compared to using `list` for `index` (e.g. `x[0].shape = (shape1, shape2)`, whereas `x[[0]].shape = (1, shape1, shape2)`. If that is the case, will that ruin our testing? Maybe we can turn the `int` into `list` as well?

File: rlscan_train/agent/buffer.py [hunk 11]
```diff
@@ -121,41 +132,42 @@ def sample_from_ind(self, indices):
...
 
-    def get_random_point(self, batch_size):
-        random_point = np.random.randint(0, self.count - batch_size)
-        while not self.done_h5[random_point] and random_point < self.count - batch_size:
-            random_point += 1
-        if not self.done_h5[random_point]:
-            while not self.done_h5[random_point] and random_point > 0:
-                random_point -= 1
-        random_point += 1
+        while not self.done_h5[random_point]:
```
Comment (chanb) [inline rlscan_train/agent/buffer.py]: I hope the memory doesn't blow up, but an alternative is use conditional indexing?
My impression is loop is usually slower... I suppose it doesn't matter for `rlscan` since the episodes are short.

File: rlscan_train/agent/policies/scanac.py [hunk 12]
```diff
@@ -11,25 +11,24 @@ class SplitBatchTrain_ScanMotionPolicy(ScanMotionPolicy_Aug27DiscreteScanAC):
...
         Returns:
             Value and log probability of the action 'a' at state 'x'
 
         """
 
-
-        x_loader = DataLoader(x, self.forward_batch_size, shuffle=False)
-        a_loader = DataLoader(a, self.forward_batch_size, shuffle=False)
+        x_loader = DataLoader(x, self.GPU_INFERENCE_BATCH_SIZE, shuffle=False)
+        a_loader = DataLoader(a.reshape(-1), self.GPU_INFERENCE_BATCH_SIZE, shuffle=False)
```
Comment (chanb) [inline rlscan_train/agent/policies/scanac.py]: Why do we want to flatten `a`?

File: rlscan_train/agent/policies/scanac.py [hunk 13]
```diff
@@ -11,25 +11,24 @@ class SplitBatchTrain_ScanMotionPolicy(ScanMotionPolicy_Aug27DiscreteScanAC):
...
 
     def val_lprob(self, x, a):
         """ Split forward pass into batches
 
         NOTE: Inputs and outputs are on cpu
         Args:
-            x: Numpy array with flattened scanner images and arm sensory information
-            a: Numpy array with the action
+            x (Torch tensor): Flattened scanner images and arm sensory information of shape [N, obs_dim]
+            a (Torch tensor): Numpy array with the actions of shape [N, acts_dim]
```
Comment (chanb) [inline rlscan_train/agent/policies/scanac.py]: Not a numpy :grin:

File: rlscan_train/agent/policies/scanac.py [hunk 14]
```diff
@@ -41,7 +40,7 @@ def val_lprob(self, x, a):
             vals = torch.cat((vals, val.cpu()))
             lprobs = torch.cat((lprobs, dist.log_prob(a_batch).unsqueeze(-1)))
```
Comment (chanb) [inline rlscan_train/agent/policies/scanac.py]: Can we just concatenate the log probs and return that, rather than `unsqueeze` and `flatten`?

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 15]
```diff
@@ -1,161 +1,225 @@
...
+        self.chunk_size = chunk_size
+        self.start_ind = opac_buffer.get_random_point_for_seq_sampler(batch_size=batch_size)
+        self.stop_ind = self.start_ind + batch_size
+        self.sampler = SubsetSampler(indices=np.arange(self.start_ind, self.stop_ind))
+        self.batch_sampler = BatchSampler(sampler=self.sampler, batch_size=chunk_size, drop_last=False)
+        self.data_loader = DataLoader(dataset=opac_buffer, shuffle=False, batch_sampler=self.batch_sampler)
+
+    def __iter__(self):
+        for experience_chunk in self.data_loader:
+            yield experience_chunk
```
Comment (chanb) [inline rlscan_train/agent/opaclearner_hdf5.py]: I personally think there should be a separate file for these classes.

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 16]
```diff
@@ -1,161 +1,225 @@
...
+        train_batch = TrainingBatchLoader(batch_size=self.cfg['batch_size'] + 1, opac_buffer=self.buffer,
+                                          chunk_size=self.RAM_CHUNK_SIZE)
+
+        # TODO: Apparently PPO pushed this outside the loop only for the sake of faster computation
+        #   We need to move this inside the for loop for correct advantage estimates
+        rews, dones, preu_vals, data_lprobs, preu_lprobs = self.torch_map_val_lprob(train_batch)
+        advs, rets = self.estimate_off_policy_advantages(rews, dones, preu_vals, data_lprobs, preu_lprobs)
+        annealer = max(1.0 - float(self.num_updates) / self.cfg['annealer_const'], 0) if self.cfg['anneal_params'] else 1
+
+        # For statistical efficiency, we re-use each data point opt_epoch times between model updates
```
Comment (chanb) [inline rlscan_train/agent/opaclearner_hdf5.py]: I'm not sure what statistical efficiency means. :thinking:

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 17]
```diff
@@ -1,161 +1,225 @@
...
+        opt_batch_sampler = BatchSampler(sampler=RandomSampler(data_source=advs),
+                                         batch_size=self.cfg['opt_batch_size'], drop_last=False)
+        for i_epoch in range(self.cfg['opt_epochs']):
+            for opt_batch_indices in opt_batch_sampler:
+                total_loss = 0
+                num_blocks = len(opt_batch_indices) // self.GPU_BLOCK_SIZE + 1
```
Comment (chanb) [inline rlscan_train/agent/opaclearner_hdf5.py]: Is it more accurate to do `math.ceil(len(opt_batch_indices)) / self.GPU_BLOCK_SIZE`?

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 18]
```diff
@@ -1,161 +1,225 @@
...
+                    gpu_block_indices.sort()
+                    obss, acts, _, _, _ = self.buffer[train_batch.start_ind + np.array(gpu_block_indices)]
+                    new_vals, new_lprobs = self.actor_critic.val_lprob(obss, acts)
+                    policy_surr = self.actor_critic_update(advs[gpu_block_indices], data_lprobs[gpu_block_indices],
+                                                           new_lprobs, preu_lprobs[gpu_block_indices], annealer,
+                                                           rets[gpu_block_indices], new_vals,
+                                                           preu_vals[gpu_block_indices])
+                    if policy_surr is None:
+                        self._logger.info("Stop aggregating gradients to prevent over-fitting")
+                        break
```
Comment (chanb) [inline rlscan_train/agent/opaclearner_hdf5.py]: I think we had a conversation about whether we should stop the update completely, or just skip the current optimizer batch. Maybe we should have a note for this.

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 19]
```diff
@@ -1,161 +1,225 @@
...
+                num_blocks = len(opt_batch_indices) // self.GPU_BLOCK_SIZE + 1
+                self.actor_critic_opt.zero_grad()
+                for gpu_block_indices in BatchSampler(sampler=SubsetSampler(indices=opt_batch_indices),
+                                                      batch_size=self.GPU_BLOCK_SIZE, drop_last=False):
+                    # NOTE: We assume that the buffer is a HDF5 dataset.
+                    # Indices must be in increasing order for HDF5 data sampling
+                    gpu_block_indices.sort()
+                    obss, acts, _, _, _ = self.buffer[train_batch.start_ind + np.array(gpu_block_indices)]
+                    new_vals, new_lprobs = self.actor_critic.val_lprob(obss, acts)
+                    policy_surr = self.actor_critic_update(advs[gpu_block_indices], data_lprobs[gpu_block_indices],
```
Comment (chanb) [inline rlscan_train/agent/opaclearner_hdf5.py]: I may be wrong, but isn't `policy_surr` just the `L_clip` loss in PPO? If so, can we clarify this loss since it's the sum of all surrogate losses?

File: rlscan_train/agent/opaclearner_hdf5.py [hunk 20]
```diff
@@ -166,71 +230,104 @@ def actor_critic_update(self, advs, data_lprobs, new_lprobs, preu_lprobs, anneal
...
 
         # Stop overfitting based on proportion of samples where clip activates
         if self.cfg['u_clip_stop']:
+            # XXX can clipping be happening on both sides?
             clip_prop = (surr2 > surr1).sum().item() / surr1.shape[0]
             if clip_prop > self.cfg['stop_const']:
                 return None
+        # XXX monitor: how many iterations did we get through, was that normal?
+
+        # XXX should this termination condition depend on the importance ratios?
```
Comment (chanb) [inline rlscan_train/agent/opaclearner_hdf5.py]: I'm curious about how annealing the clipping value affects your questions above.

File: rlscan_train/agent/buffer.py (hunk 10 above)
Comment (gauthamvasan) [reply rlscan_train/agent/buffer.py]: You are right. `(1, shape1, shape2)` helps stack arrays later in the code. It's handled in testing as well. Maybe just make it obvious in comments?

File: rlscan_train/agent/buffer.py (hunk 11 above)
Comment (gauthamvasan) [reply rlscan_train/agent/buffer.py]: Conditional indexing is the better solution. Again, we run rlscan based on certain assumptions

File: rlscan_train/agent/buffer.py (hunk 10 above)
Comment (chanb) [reply rlscan_train/agent/buffer.py]: I think it may be better if we make the dimensions consistent regardless of input type

File: rlscan_train/agent/opaclearner_hdf5.py (hunk 19 above)
Comment (gauthamvasan) [reply rlscan_train/agent/opaclearner_hdf5.py]: This policy_surr is a combination of losses. The name is a bit misleading. Will switch to calling it loss.

File: rlscan_train/agent/policies/scanac.py (hunk 12 above)
Comment (gauthamvasan) [reply rlscan_train/agent/policies/scanac.py]: Reshape from [N, 1] to [N] for inference pass

File: rlscan_train/agent/opaclearner_hdf5.py (hunk 20 above)
Comment (gauthamvasan) [reply rlscan_train/agent/opaclearner_hdf5.py]: Something that needs further experimentation :)

File: rlscan_train/agent/opaclearner_hdf5.py (hunk 16 above)
Comment (gauthamvasan) [reply rlscan_train/agent/opaclearner_hdf5.py]: Another way of saying the same would be sample efficiency I guess. We're trying to get as much useful gradient info as possible from the generated data.

File: rlscan_train/agent/policies/scanac.py (hunk 12 above)
Comment (chanb) [reply rlscan_train/agent/policies/scanac.py]: I see. Actions are scalars in this case right? I thought it was a vector (This is converted later I suppose)
